plugins/pulp_docker/plugins/importers/upload.py

```diff
@@ -169,6 +179,47 @@ def get_models(self, metadata, mask_id=''):
         return images
 
 
+class ImportManifestList(PluginStep):
+    """
+    Add an uploaded ManifestList to a repository.
+    """
+
+    def process_main(self):
+        """
+        Validate the uploaded manifest list json, then import content unit into repository.
+        """
+        with open(self.parent.file_path, 'r') as uploaded_file:
+            manifest_list = uploaded_file.read()
+        models.ManifestList.validate_json(manifest_list)
+        digest = models.Manifest.calculate_digest(manifest_list)
+        manifest_list_instance = models.ManifestList.from_json(manifest_list, digest)
+        transfer_repo = self.get_repo()
+
+        #  Ensure that all referenced manifests are already in repository.
+        manifest_digests = set()
+        for manifest in manifest_list_instance.manifests:
+            manifest_digests.add(manifest)
+        qs = models.Manifest.objects.filter(
+            digest__in=sorted(manifest_digests)).only('id', 'digest')
+        referenced_manifests = {manifest['digest']: manifest['id'] for manifest in qs}
+        unit_qs = pulp_models.RepositoryContentUnit.objects.filter(
+            repo_id=transfer_repo.id,
+            unit_id__in=referenced_manifests.values()).values_list('unit_id')
+        unit_ids_in_repo = list(unit_qs)
+        for digest in manifest_digests:
+            unit_id = referenced_manifests.get(digest)
+            if not unit_id or unit_id not in unit_ids_in_repo:
+                raise PulpCodedValidationException(error_code=error_codes.DKR1013,
+                                                   digest=digest)
+
+        try:
+            manifest_list_instance.save_and_import_content(self.parent.file_path, digest)
+        except NotUniqueError:
+            raise PulpCodedValidationException(error_code=pulp_error_codes.PLP0018,
```

Comment:
i don't think we want to raise here the dup resource error.
this will fail in case you'd want to upload  manifest-list A into a repository B but it is also present in repo A
also there is something weird happening with the digest.
```
$ pulp-admin  docker repo search manifest-list --repo-id containers --str-eq='digest=sha256:03090180de84b192027dfc77517f32cdec2716d5c3ef4703b5cda09457e3cb7d'
Warning: path should have mode 0700 because it may contain sensitive information: /home/ipanova/.pulp/

Created:      2017-09-21T15:31:57Z
Metadata:     
  Amd64 Digest:         sha256:4831b0e3611d831c0bb5c9e44b0a538adfacb4ec3cab0ff20
                        77423ce37273055
  Amd64 Schema Version: 2
  Digest:               sha256:03090180de84b192027dfc77517f32cdec2716d5c3ef4703b
                        5cda09457e3cb7d
  Downloaded:           True
  Manifests:            sha256:4831b0e3611d831c0bb5c9e44b0a538adfacb4ec3cab0ff20
                        77423ce37273055,
                        sha256:b5dbad4bdb4444d919294afe49a095c23e86782f98cdf0aa2
                        86198ddb814b50b,
                        sha256:6fa7cb96c7d16130cf81f8404da04e53b7bb589261de420ab
                        8695b0cb6b0a0b7,
                        sha256:96d81636bee8ff999651faaa380845ad8affd4509d5836de4
                        fe9a5d06c3ed0c0,
                        sha256:a30d01ad10895e27b65dc284ae9c493c968e0bbce2cefd112
                        631e8b2672a7927,
                        sha256:03bef693cdd911df35642c83e371faf64d3a362a0dfeb5bc0
                        95de2408c06f1d5
  Pulp User Metadata:   
  Schema Version:       2
Repo Id:      containers
Unit Id:      6c76881f-1b87-4513-874c-5295cc287027
Unit Type Id: docker_manifest_list
Updated:      2017-09-21T15:31:59Z


[ipanova@ina pulp_docker]$ pulp-admin  docker repo uploads upload --repo-id containers2 --file /home/ipanova/pulp_development/pulp_docker/sha256:03090180de84b192027dfc77517f32cdec2716d5c3ef4703b5cda09457e3cb7d
Warning: path should have mode 0700 because it may contain sensitive information: /home/ipanova/.pulp/

+----------------------------------------------------------------------+
                              Unit Upload
+----------------------------------------------------------------------+

Extracting necessary metadata for each request...
[==================================================] 100%
Analyzing:
sha256:03090180de84b192027dfc77517f32cdec2716d5c3ef4703b5cda09457e3cb7d
... completed

Creating upload requests on the server...
[==================================================] 100%
Initializing:
sha256:03090180de84b192027dfc77517f32cdec2716d5c3ef4703b5cda09457e3cb7d
... completed

Starting upload of selected units. If this process is stopped through ctrl+c,
the uploads will be paused and may be resumed later using the resume command or
canceled entirely using the cancel command.

Uploading:
sha256:03090180de84b192027dfc77517f32cdec2716d5c3ef4703b5cda09457e3cb7d
[==================================================] 100%
2029/2029 bytes
... completed

Importing into the repository...
This command may be exited via ctrl+c without affecting the request.


[\]
Running...

Task Failed

Duplicate resource:
sha256:6fa7cb96c7d16130cf81f8404da04e53b7bb589261de420ab8695b0cb6b0a0b7

Deleting the upload request...
... completed
```

Reply:
actually this is what is going on with the digest, it is the image manifest from this loop https://github.com/pulp/pulp_docker/pull/209/files#diff-e77f69fbead8c3055ea1da0fef1f6132R209 and not the manifest list digest

Reply:
I fixed the digest issue. Also, I've updated it so that if the uploaded file is a duplicate, the original db entry is retrieved and associated to the repository. This means we will no longer fail, even if the user submits the same file to be uploaded and associated twice.

Reply:
this is what is expected, but let's stick to the pattern we use everywhere. example https://github.com/pulp/pulp_docker/blob/master/plugins/pulp_docker/plugins/importers/upload.py#L348